doc/03-omnifest/02-external.md

```diff
@@ -1,22 +1,117 @@
 # External
 
-## osbuild
+External directives are directives that are implemented externally from `otk`.
+They are meant to be used to provide target-specific behavior and can be used
+to express things where `otk` is not expressive enough.
+
+## Protocol
+
+Directives are small executables that receive JSON and are expected to output
+JSON again in a specific format. Additional keys are not allowed.
+
+When we have an [omnifest](./index.md) that looks like this:
+
+```yaml
+otk.external.name:
+  child:
+    - 1
+  options: "are here"
+```
+
+The external gets called with the following JSON:
+
+```json
+{
+  "tree": {
+    "otk.external.name": {
+      "child": [1],
+      "options": "are here"
+    }
+  }
+}
+```
+
+And is expected to return the following JSON:
+
+```json
+{
+  "tree": {
+    "what": {
+      "you": "want"
+    }
+  }
+}
+```
+
+Which will lead to the following YAML structure:
+
+```yaml
+tree:
+  what:
+    you: "want"
+```
+
+If the external returns `{}`, an empty object, `otk` will assume that there is
+no tree to replace and remove the node instead. This will turn the following
+YAML:
+
+```yaml
+dummy:
+  otk.external.name:
+    options:
+```
+
+Into this YAML structure:
+
+```yaml
+dummy:
+```
+
```

Comment:
Perhaps a "hello world" style example would be nice here too, something that's fully working but doesn't do anything interesting.

I'm thinking something like:

## Example

The following example demonstrates how an external can be used to implement simple string concatenation.

Given the following script, called `concat`, in `/usr/local/libexec/otk/concat`
```bash
#!/usr/bin/env bash

output="$(jq -jr '.tree."otk.external.concat".parts[]' <<< "${1}")"
echo "{\"tree\":{\"output\":\"$output\"}}"
```

and the following directive:
```yaml
examplestring:
  otk.external.concat:
    parts:
      - list
      - of
      - strings
```

the script is called with the following input:
```json
{
  "tree": {
    "otk.external.concat": {
      "parts": [
        "list",
        "of",
        "strings"
      ]
    }
  }
}
```

which results in the following output:
```json
{
  "tree": {
    "string": "listofstrings"
  }
}
```

and the final omnifest will be:
```yaml
examplestring:
  listofstrings
```

Reply:
This is now included in the PR.